src/AggregateFunctions/IAggregateFunction.h

```diff
@@ -97,6 +97,12 @@ class IAggregateFunction
     /// Deserializes state. This function is called only for empty (just created) states.
     virtual void deserialize(AggregateDataPtr place, ReadBuffer & buf, Arena * arena) const = 0;
 
+    /// Finalize state. This function is called once after all 'add' and 'merge' calls. Only if isFinalizationNeeded().
+    /// 'add', 'merge', and 'serialize' can't be performed after 'finalize' call.
```

Comment:
If we can't serialize the finalized state, this means it would be a mistake to send a block with a finalized ColumnAggregateFunction over network? Is there some check to guarantee we won't do this?

Reply:
It would be nice to add, though I don't know how. We can't just check it for every state.
The best way I see is to support invariants in `Aggregator` and `ColumnAggregateFunction` interfaces:
* `Aggregator` can return column with already calculated values or  `ColumnAggregateFunction`. There is no way to get raw aggregating data.
* `ColumnAggregateFunction` always stores not finalized aggregation states.
* We can only once call `ColumnAggregateFunction::convertToValues`. After that states will be finalized only once, and `ColumnAggregateFunction` shouldn't be used any more. Let's make `convertToValues` rvalue-qualified.